### PR TITLE
feat: introduce `Range` component

### LINKED ISSUE #376: introduce `Psl\Range` component
```php
$a = Range\from(2); // 2..
$b = Range\to(10, inclusive: true); // ..=10
$c = Range\between(1, 20, inclusive: true); // 1..=20
$d = Range\to(10, inclusive: false); // ..10
$e = Range\between(1, 20, inclusive: false); // 1..20
$f = Range\full(); // ..
```

we would introduce the following classes:

- `Psl\Range\ToRange`
- `Psl\Range\FromRange`
- `Psl\Range\BetweenRange`
- `Psl\Range\FullRange`
- `Psl\Range\LowerBoundInterface` includes `getLowerBound(): int` method
- `Psl\Range\UpperBoundInterface` includes `getUpperBound(): int` method, and `isInclusive(): bool` method.

`BetweenRange`, `ToRange`, and `FromRange` would implement `UpperBoundInterface` interface.
`BetweenRange` and `FromRange` would implement `LowerBoundInterface` interface.

`BetweenRange` and `FromRange` would implement `IteratorAggregate` interface.

`FromRange::getUpperBound()` would return infinity.

iterating over `FromBound` would iterate from the lower bound to infinity

### PR DESCRIPTION
**Pull request by azjezz**
closes #376 

---

```php
// 0..

$range = Range\from(0);

$range->contains(-100); // false
$range->contains(0); // true
$range->contains(99); // true
$range->contains(100); // true
$range->getLowerBound(); // 0
$range->getIterator();   // 0..infinity

// ..=100

$range = Range\to(100, upper_inclusive: true);

$range->contains(-100); // true
$range->contains(0); // true
$range->contains(99); // true
$range->contains(100); // true
$range->contains(101); // false
$range->isUpperInclusive(); // true
$range->getUpperBound(); // 100

// ..100

$range = Range\to(100, upper_inclusive: false);

$range->contains(-100); // true
$range->contains(0); // true
$range->contains(99); // true
$range->contains(100); // false
$range->contains(101); // false
$range->isUpperInclusive(); // false
$range->getUpperBound(); // 100

// 0..=100

$range = Range\between(0, 100, upper_inclusive: true);

$range->contains(0); // true
$range->contains(99); // true
$range->contains(100); // true
$range->isUpperInclusive(); // true
$range->getLowerBound(); // 0
$range->getUpperBound(); // 100
$range->getIterator(); // 0..=100

// 0..100

$range = Range\between(0, 100, upper_inclusive: false);

$range->contains(0); // true
$range->contains(99); // true
$range->contains(100); // false
$range->isUpperInclusive(); // false
$range->getLowerBound(); // 0
$range->getUpperBound(); // 100
$range->getIterator(); // 0..100

// ..

$range = Range\full();
$range->contains(100000); // always true
```

### CODE REVIEW

File: src/Psl/Range/UpperBoundRangeInterface.php [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+     * @psalm-mutation-free
+     */
+    public function getUpperBound(): int|float;
+
+    /**
+     * Returns whether the upper bound is inclusive.
+     *
+     * @psalm-mutation-free
+     */
+    public function isInclusive(): bool;
```
Comment (azjezz) [inline src/Psl/Range/UpperBoundRangeInterface.php]: this should be `isUpperInclusive()`

File: src/Psl/Range/between.php [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+ * @template T of int|float
+ *
+ * @param T $lower_bound
+ * @param T $upper_bound
+ *
+ * @return BetweenRange<T>
+ *
+ * @psalm-mutation-free
+ */
+function between(int|float $lower_bound, int|float $upper_bound, bool $inclusive = false): BetweenRange
```
Comment (azjezz) [inline src/Psl/Range/between.php]:
```suggestion
function between(int|float $lower_bound, int|float $upper_bound, bool $upper_inclusive = false): BetweenRange
```

File: src/Psl/Range/to.php [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+/**
+ * @template T of int|float
+ *
+ * @param T $upper_bound
+ *
+ * @return ToRange<T>
+ *
+ * @psalm-mutation-free
+ */
+function to(int|float $upper_bound, bool $inclusive = false): ToRange
```
Comment (azjezz) [inline src/Psl/Range/to.php:16]:
```suggestion
function to(int|float $upper_bound, bool $upper_inclusive = false): ToRange
```

File: src/Psl/Range/BetweenRange.php [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    /**
+     * @param T $lower_bound
+     * @param T $upper_bound
+     *
+     * @psalm-mutation-free
+     */
+    public function __construct(
+        private readonly int|float $lower_bound,
+        private readonly int|float $upper_bound,
+        private readonly bool $inclusive = false,
```
Comment (azjezz) [inline src/Psl/Range/BetweenRange.php]:
```suggestion
        private readonly bool $upper_inclusive = false,
```

File: src/Psl/Range/ToRange.php [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+final class ToRange implements UpperBoundRangeInterface
+{
+    /**
+     * @param T $upper_bound
+     *
+     * @psalm-mutation-free
+     */
+    public function __construct(
+        private readonly int|float $upper_bound,
+        private readonly bool $inclusive = false,
```
Comment (azjezz) [inline src/Psl/Range/ToRange.php]:
```suggestion
        private readonly bool $upper_inclusive = false,
```

Comment (azjezz) [reply src/Psl/Range/ToRange.php]: https://github.com/azjezz/psl/compare/90789472bfbfcb453e6fa4d6303b5905e052e396..c99e6ab756e79c98d41deda361a4cf5d3a400d3b

File: src/Psl/Range/BetweenRange.php (hunk 4 above)
Comment (azjezz) [reply src/Psl/Range/BetweenRange.php]: https://github.com/azjezz/psl/compare/90789472bfbfcb453e6fa4d6303b5905e052e396..c99e6ab756e79c98d41deda361a4cf5d3a400d3b

File: src/Psl/Range/to.php (hunk 3 above)
Comment (azjezz) [reply src/Psl/Range/to.php:16]: https://github.com/azjezz/psl/compare/90789472bfbfcb453e6fa4d6303b5905e052e396..c99e6ab756e79c98d41deda361a4cf5d3a400d3b

File: src/Psl/Range/between.php (hunk 2 above)
Comment (azjezz) [reply src/Psl/Range/between.php]: https://github.com/azjezz/psl/compare/90789472bfbfcb453e6fa4d6303b5905e052e396..c99e6ab756e79c98d41deda361a4cf5d3a400d3b

File: src/Psl/Range/UpperBoundRangeInterface.php (hunk 1 above)
Comment (azjezz) [reply src/Psl/Range/UpperBoundRangeInterface.php]: https://github.com/azjezz/psl/compare/90789472bfbfcb453e6fa4d6303b5905e052e396..c99e6ab756e79c98d41deda361a4cf5d3a400d3b

**Comment by azjezz** [comment]
for a future PR, we can add `LowerBoundRangeInterface::to(T $upper, bool $upper_inclusive): LowerBoundRangeInterface&UpperBoundRangeInterface`, so you can do:

```php
$x = Range\from(10)->to(100, false);
```

which would be the same as:

```php
$x = Range\between(10, 100, false);
```

**Review by veewee — APPROVED** [review]
LGTM 🚢

**Pull request merged**